index.js

```diff
@@ -1,77 +1,184 @@
+
+    if (this.unwatch) {
+      await this.unwatch()
+      this.unwatch = null
     }
-    clear = st.isDirectory() ? watchDirectory(name, onchange) : watchFile(name, onchange)
-  })
+  }
+}
+
+class Watch {
+  constructor (filename, opts, onchange) {
+    this.filename = filename
+    this.opts = opts || {}
+    this.onchange = onchange
+
+    this.opened = true
+    this.closed = false
 
-  return function () {
-    if (stopped) return
-    stopped = true
-    if (clear) {
-      clear()
+    this._closing = null
+
+    this.watcher = fs.watch(this.filename, this.opts, this.onchange)
+    this.watcherClosed = false
+
+    this.watcher.on('error', noop)
+    this.watcher.on('close', () => {
+      this.watcherClosed = true
+    })
+  }
+
+  async close () {
+    if (this._closing) return this._closing
+    this._closing = this._close()
+    return this._closing
+  }
+
+  async _close () {
+    if (this.closed) return
+    this.closed = true
+
+    let onclose = null
+
+    if (!this.watcherClosed) {
+      onclose = new Promise(resolve => this.watcher.on('close', resolve))
     }
+
+    this.watcher.close()
+    await onclose
+
+    this.watcher = null
   }
 }
 
 function watchFile (filename, onchange) {
-  var prev = null
-  var prevTime = 0
+  let prev = null
+  let prevTime = 0
+  let actives = 0
+
+  let closed = false
+  let close = null
+  const closing = new Promise(resolve => { close = resolve })
+
+  const watcher = new Watch(filename, null, function () {
+    actives++
 
-  var w = fs.watch(filename, function () {
     fs.lstat(filename, function (_, st) {
-      var now = Date.now()
+      actives--
+
+      if (closed) {
+        oncleanup()
+        return
+      }
+
+      const now = Date.now()
       if (now - prevTime > 2000 || !same(st, prev)) onchange(filename)
       prevTime = now
       prev = st
     })
   })
 
-  return function () {
-    w.close()
+  return async function unwatch () {
+    if (closed) return closing
+    closed = true
+
+    await watcher.close()
+    oncleanup()
+
+    return closing
+  }
+
+  function oncleanup () {
+    if (actives === 0) close()
   }
 }
 
 function watchRecursive (directory, onchange) {
-  var w = fs.watch(directory, {recursive: true}, function (change, filename) {
-    if (!filename) return // filename not always given (https://nodejs.org/api/fs.html#fs_filename_argument)
+  const watcher = new Watch(directory, { recursive: true }, function (change, filename) {
+    if (!filename) return // Filename not always given by fs.watch
+
     onchange(path.join(directory, filename))
   })
 
-  return function () {
-    w.close()
+  return function unwatch () {
+    return watcher.close()
   }
 }
 
 function watchFallback (directory, onchange) {
-  var watching = {}
-  var loaded = false
-  var queued = []
-  var prevs = new Cache({ttl: 2e3, capacity: 30})
+  let loaded = false
+  const queued = []
+  const prevs = new Xache({ maxSize: 30, maxAge: 2000 })
```

Comment:
This 2000 refers to the same 2000 as used in here I suppose? 
```
    fs.lstat(filename, function (_, st) {
      actives--

      if (closed) {
        oncleanup()
        return
      }

      const now = Date.now()
      if (now - prevTime > 2000 || !same(st, prev)) onchange(filename)
      prevTime = now
      prev = st
    })
  })
```

If so, I'd extract it as a constant, I think the code will break if they ever get out of sync

Reply:
Agreed on the constant